runtime/obj.c

```diff
@@ -90,16 +97,16 @@ CAMLprim value caml_obj_block(value tag, value size)
 }
 
 /* Spacetime profiling assumes that this function is only called from OCaml. */
-CAMLprim value caml_obj_dup(value arg)
+CAMLprim value caml_obj_with_tag(value new_tag_v, value arg)
 {
-  CAMLparam1 (arg);
+  CAMLparam2 (new_tag_v, arg);
   CAMLlocal1 (res);
   mlsize_t sz, i;
   tag_t tg;
 
   sz = Wosize_val(arg);
+  tg = (tag_t)Long_val(new_tag_v);
   if (sz == 0) CAMLreturn (arg);
```

Comment:
I'm not sure that this `sz == 0` optimization is correct if the tag is different.

Reply:
Wow, that's clearly wrong. Nice catch.